src/tests/testClpMap.cc

```diff
@@ -153,55 +165,41 @@ testClpMap::testConstructor()
 }
 
 void
-testClpMap::testSetMemLimit()
+testClpMap::testMemoryLimit()
 {
-    TestMap m(2048);
-    // overflow the map with entries to make sure it has lots of entries to purge below
-    fillMapWithElements(m);
-    auto testEntriesBefore = m.entries();
-    CPPUNIT_ASSERT(testEntriesBefore > 0);
-
-    m.setMemLimit(m.memoryUsed() / 2);
-    auto entriesAfterPurge = m.entries();
-    CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
-
-    m.setMemLimit(m.memLimit() * 2);
-    // overflow the map with entries again to make sure it can grow after purging
+    const size_t initialCapacity = 1024; // bytes
+    TestMap m(initialCapacity);
     fillMapWithElements(m);
-    CPPUNIT_ASSERT(entriesAfterPurge < m.entries());
-
-    // progressively shrink to size 0, checking that the number of
-    // entries decreases each time
-    testEntriesBefore = m.entries();
-    const auto singleEntrySizeEstimate = *TestMap::MemoryCountedFor("0", 0) + 10;
-    while (m.memoryUsed() > singleEntrySizeEstimate) {
-        m.setMemLimit(m.memoryUsed() / 2);
-        entriesAfterPurge = m.entries();
-        CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
-        CPPUNIT_ASSERT(entriesAfterPurge > 0);
-        testEntriesBefore = entriesAfterPurge;
-    }
+    const auto entriesAtInitialCapacity = m.entries();
 
-    // set size to 0, all entries should have been removed
+    // check that all entries are removed if we prohibit storage of any entries
     m.setMemLimit(0);
     CPPUNIT_ASSERT_EQUAL(size_t(0), m.entries());
 
-    // test that memory used decreases when shrinking
-    size_t curLimit = 10240;
-    size_t lastMemoryUsed;
-    m.setMemLimit(curLimit);
+    // test whether the map can grow after the all-at-once purging above
+    m.setMemLimit(initialCapacity * 2);
     fillMapWithElements(m);
-    lastMemoryUsed = m.memoryUsed();
-    while (curLimit > 1024) { // stop at any practical size
-        CPPUNIT_ASSERT(m.memoryUsed() < curLimit);
-        CPPUNIT_ASSERT(m.memoryUsed() <= lastMemoryUsed);
-        // also check that we can still add entries - evicting old ones if needed
-        CPPUNIT_ASSERT(m.add(std::to_string(curLimit+1000), 1, 10));
-        CPPUNIT_ASSERT_EQUAL(1, *m.get(std::to_string(curLimit + 1000)));
-        curLimit /= 2;
-        lastMemoryUsed = m.memoryUsed();
-        m.setMemLimit(curLimit);
+    CPPUNIT_ASSERT(m.entries() > entriesAtInitialCapacity);
+
+    // test that memory usage and entry count decrease when the map is shrinking
+    while (m.entries()) {
```

Comment:
I'm a bit skeptical about this test condition: if there were a bug in ClpMap and entries didn't shrink, this would cause the test to endlessly loop; not the clearest outcome, and unfriendly to automations

Reply:
This loop cannot run endlessly because it asserts that an unsigned value (m.entries()) always decreases.

Reply:
I think I wasn't clear. What I am considering is a much simpler:

```suggestion
    int maxIterations = 10000;
    while (m.entries() && --maxIterations > 0) {
```

Reply:
TLDR: I added a source code comment in hope to better address your concern (commit d494db2c3463195a4c4565ba5760140be996e805).

----

Original change request: This loop may be endless if ClpMap is buggy.

Correction: This loop is not endless even if ClpMap is buggy.

Modified change request: Let's consider adding more code, adding more implicit dependencies among constants, and making the loop condition more complex.

I assume that all those changes are meant to make it more obvious that the loop is not endless. IMO, that goal is just not worth the cost, especially when the worst thing that could happen[^1] (with or without) those changes is an endless loop during "make check".

[^1]: And only in a scenario where somebody breaks the testing code itself -- something no solution in this testing code, not even the proposed explicit counter, can protect us from! Again, breaking ClpMap will _not_ break this testing code.

Personally, I would have _more_ questions about the loop with maxIterations code than without it. And I can easily imagine other developers starting to copy (or demand) similar paranoid additions to similar loops in new or modified code!

However, this is _your_ PR. If you insist on adding that code, then do it. My only request is that you also add a comment to explain why that extra code is there. I suggested specific wording in your other change request (because we need to explain the new proposed assertion anyway).

Reply:
This thread has been switched to https://github.com/squid-cache/squid/pull/1262/files#r1116614481